renderdoc/driver/vulkan/vk_acceleration_structure.cpp

```diff
@@ -450,320 +618,261 @@ void VulkanAccelerationStructureManager::CopyAccelerationStructure(
+    const VkMappedMemoryRange range = {
+        VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE, NULL, geomData.readbackMem, 0, geomData.memSize,
+    };
 
-  // invalidate the cpu cache for this memory range to avoid reading stale data
-  const VkMappedMemoryRange range = {
-      VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE, NULL, Unwrap(readbackmem.mem), readbackmem.offs, size,
-  };
-  vkr = ObjDisp(d)->InvalidateMappedMemoryRanges(Unwrap(d), 1, &range);
-  CHECK_VKR(m_pDriver, vkr);
+    vkr = ObjDisp(d)->InvalidateMappedMemoryRanges(Unwrap(d), 1, &range);
+    CHECK_VKR(m_pDriver, vkr);
 
-  // Count the BLAS device addresses to update the AS type
-  const uint64_t handleCount = *(uint64_t *)(mappedDstBuffer + handleCountOffset);
-  result = {readbackmem, true};
-  result.isTLAS = handleCount > 0;
+    ser.Serialise("AS Input"_lit, contents, geomData.memSize, SerialiserFlags::NoFlags).Hidden();
 
-  ObjDisp(d)->UnmapMemory(Unwrap(d), Unwrap(result.alloc.mem));
+    ObjDisp(d)->UnmapMemory(Unwrap(d), geomData.readbackMem);
+  }
 
   return true;
 }
 
-template <typename SerialiserType>
-bool VulkanAccelerationStructureManager::Serialise(SerialiserType &ser, ResourceId id,
+bool VulkanAccelerationStructureManager::Serialise(ReadSerialiser &ser, ResourceId id,
                                                    const VkInitialContents *initial,
                                                    CaptureState state)
 {
-  VkDevice d = !IsStructuredExporting(state) ? m_pDriver->GetDev() : VK_NULL_HANDLE;
-  const bool replayingAndReading = ser.IsReading() && IsReplayMode(state);
-  VkResult vkr = VK_SUCCESS;
-
-  byte *contents = NULL;
-  uint64_t contentsSize = initial ? initial->mem.size : 0;
-  MemoryAllocation mappedMem;
-
-  // Serialise this separately so that it can be used on reading to prepare the upload memory
-  SERIALISE_ELEMENT(contentsSize);
-
   const VkDeviceSize nonCoherentAtomSize = m_pDriver->GetDeviceProps().limits.nonCoherentAtomSize;
 
-  // the memory/buffer that we allocated on read, to upload the initial contents.
-  MemoryAllocation uploadMemory;
-  VkBuffer uploadBuf = VK_NULL_HANDLE;
+  VkAccelerationStructureInfo *asInfo = new VkAccelerationStructureInfo();
+  SERIALISE_ELEMENT(*asInfo).Hidden();
 
-  if(ser.IsWriting())
+  VkDevice d = !IsStructuredExporting(state) ? m_pDriver->GetDev() : VK_NULL_HANDLE;
+  VkResult vkr = VK_SUCCESS;
+
+  for(VkAccelerationStructureInfo::GeometryData &geomData : asInfo->geometryData)
   {
-    if(initial && initial->mem.mem != VK_NULL_HANDLE)
+    Allocation uploadMemory;
+    byte *contents = NULL;
+
+    if(IsReplayMode(state) && !ser.IsErrored())
     {
-      const VkDeviceSize size = AlignUp(initial->mem.size, nonCoherentAtomSize);
+      uploadMemory =
+          CreateReplayMemory(MemoryType::Upload, geomData.memSize,
+                             VK_BUFFER_USAGE_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT_KHR);
```

Comment:
The upload memory does not need this build input usage flag.

Reply:
GPU memory is usually the limiting factor on mobile so I just went with one copy, if performance is the bigger concern on desktop then I'll switch to that and maybe keep this path in our fork.

Reply:
We only need to hold onto the GPU-local buffer for a short period of time so yes please change this to take the performant path.